APE17.rst

```diff
@@ -0,0 +1,482 @@
+after branching, e.g. after creating a ``v4.0.x`` branch we should tag the next
+commit on ``master`` as ``v4.1.dev``.
+
+Package data and entry points
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At the moment, package data and entry points for packages can be defined via
+``get_package_data`` and ``get_entry_points`` in ``setup_package.py`` files.
+However, this adds unnecessary complexity, as even for the core package it is
+simple to define the data and entry points in ``setup.cfg`` using the
+``[options.entry_points]`` and ``[options.package_data]`` sections, e.g.::
+
+    [options.entry_points]
+    console_scripts =
+        fits2bitmap = astropy.visualization.scripts.fits2bitmap:main
+    ...
+
+    [options.package_data]
+    astropy = astropy.cfg, CITATION, **/data/**/*
+    ...
+
+Removing the ability to specify package data in `setup_package.py` files removes
+the dependence of the `setup.py sdist` command on astropy-helpers, which is
+essential to using the PEP 518 build time dependencies discussed below.
+
+Using ``setup.cfg`` to define package data and other options does rely on
+setuptools 30.3 or later, which is now over two years old. Nevertheless, to
+minimize issues for users with older Python installations, we recommend
+including a version check for setuptools inside the ``setup.py`` file.
+
+Removal of pre/post-processing hooks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A little-known feature of astropy-helpers is the ability to define hooks in
+``setup_package.py`` files for steps to be carried out before/after specific
+``setup.py`` commands. We propose removing this functionality since this feature
+was never well advertised and likely only used in the core astropy package, and
+our proof-of-concept implementation in the core package shows that we can easily
+remove this.
+
+Changes for packages with compiled extensions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Build-time dependencies
+~~~~~~~~~~~~~~~~~~~~~~~
+
+One of the common issues that packages have regularly run into and which `PEP
+518`_ solves is how to define dependencies required for building a package. The
+``setup()`` function provided by setuptools takes a ``setup_requires`` argument
+which can include dependencies for the build, but unfortunately the ``setup.py``
+file has to be executed before the ``setup()`` function is run, which leads to
+the circular issue that any code in ``setup.py`` can’t rely on packages
+specified in ``setup_requires`` - thus, ``setup_requires`` does not properly
+solve the issue of build-time dependencies since those dependencies are only
+installed part way through the build process.
+
+PEP 518 instead specifies that build-time dependencies can be specified in a
+file called ``pyproject.toml`` that contains for example::
+
+    [build-system]
+    requires = ["setuptools", "wheel", "numpy==1.13.3"]
+
+Provided that the package is installed with a tool such as `pip`_ which
```

Comment:
Looks like you haven't defined the reference for this pip link.

Reply:
Fixed